### PR TITLE
Remove persistant callbacks

### PR DESCRIPTION
**Pull request by pan-**
This pull request improve `CallChainOfFunctionPointersWithContext`, `FunctionPointerWithContext` and provide a way to unregister callbacks previously registered in a callchain.

Changes in `FunctionPointerWithContext` are not minor, goals were to provide function pointer semantic and value semantic: 
  - support of copy constructor and assignement
  - support of equality operator
  - support of function call operator
  - support of safe comparison to bool
  - call does *not* call the `next` `FunctionPointerWithContext`, it is not its responsibility any more. This change can be seen as a breaking one. `FunctionPointerWithContext` needs to handle less things, is it a list, a node of a list, a call chain ? Unfortunately, mixing responsibilities hurt and I was forced to make this change if I wanted to make callback called only once worked.

`CallChainOfFunctionPointersWithContext` improvements: 
  - support some semantic of function pointer (comparison to bool and operator call). 
  - Provide detach operation, this operation remove a function pointer from the callchain, it is also allowed to remove a FunctionPointer while it is called. 
  - Handle calls of the callchain, it is not recursive anymore and it support the removal of any nodes while the callchain is traversed.

`DiscoveredCharacteristic` improvements:
  - read operation support continuation (callback)
  - write operation support continuation (callback)

`Gap` improvements: 
  - multiple timeout event handlers can be attached and detached
  - multiple connection event handlers can be attached and detached
  - multiple disconnection event handlers can be attached and detached

`GattClient` improvements:
  - multiple read event handlers can be attached and detached
  - multiple write event handlers can be attached and detached
  - multiple HVX event handlers can be attached and detached

`GattServer` improvements:
  - multiple data sent event handlers can be attached and detached
  - multiple data read event handlers can be attached and detached
  - multiple data written event handlers can be attached and detached
  - `EventCallback_t` is now a `FunctionPointerwithContext` instead of a plain function pointer

`ServiceDiscovery` callbacks are FunctionPointerWithcontext instances instead of plain function pointer

### CODE REVIEW

File: ble/CallChainOfFunctionPointersWithContext.h [hunk 1]
```diff
@@ -97,6 +97,47 @@ class CallChainOfFunctionPointersWithContext {
...
+    void add(const FunctionPointerWithContext<ContextType>& func) {
+        common_add(new FunctionPointerWithContext<ContextType>(func));
+    }
+
+    /** 
+     * Detach a function pointer from a callchain
+     * 
+     * @oaram toDetach FunctionPointerWithContext to detach from this callchain
+     * 
+     * @return true if a function pointer has been detached and false otherwise
```
Comment (rgrover) [inline ble/CallChainOfFunctionPointersWithContext.h]: it would be nice to return true if detach was successful. currently the doc-header doesn't match the function.

File: ble/CallChainOfFunctionPointersWithContext.h [hunk 2]
```diff
@@ -120,11 +161,40 @@ class CallChainOfFunctionPointersWithContext {
      *        chained FunctionPointers.
      */
     void call(ContextType context) {
```
Comment (rgrover) [inline ble/CallChainOfFunctionPointersWithContext.h]: the note (above) about all calls stacking up is no longer true.

File: ble/CallChainOfFunctionPointersWithContext.h [hunk 3]
```diff
@@ -120,11 +161,40 @@ class CallChainOfFunctionPointersWithContext {
...
     }
 
+    /**
+     * @brief same as above but with function call operator
+     */
+    void operator()(ContextType context) const {
+        call(context);
+    }
+
+    typedef void (CallChainOfFunctionPointersWithContext::*bool_type)() const;
```
Comment (rgrover) [inline ble/CallChainOfFunctionPointersWithContext.h]: why is this needed?

File: ble/DiscoveredCharacteristic.h [hunk 4]
```diff
@@ -83,6 +83,8 @@ class DiscoveredCharacteristic {
      */
     ble_error_t read(uint16_t offset = 0) const;
 
+    ble_error_t read(uint16_t offset, const GattClient::ReadCallback_t& onRead) const;
```
Comment (rgrover) [inline ble/DiscoveredCharacteristic.h]: I'm not sure this is the best API pattern. I accept that we need a way to setup callbacks for read, but it could also be `discoveredChar.read(offset).onRead(callback);`

Comment (rgrover) [inline ble/DiscoveredCharacteristic.h]: we need a discussion around this before going ahead.

File: ble/DiscoveredCharacteristic.h [hunk 5]
```diff
@@ -135,6 +137,11 @@ class DiscoveredCharacteristic {
      */
     ble_error_t write(uint16_t length, const uint8_t *value) const;
 
+    /** 
+     * Same as above but register the callback wich will be called once the data has been written
+     */
+    ble_error_t write(uint16_t length, const uint8_t *value, const GattClient::WriteCallback_t& onRead) const;
```
Comment (rgrover) [inline ble/DiscoveredCharacteristic.h]: same as above. needs discussion.

File: ble/CallChainOfFunctionPointersWithContext.h [hunk 6]
```diff
@@ -120,11 +161,40 @@ class CallChainOfFunctionPointersWithContext {
...
+    }
+
+    /**
+     * @brief same as above but const 
+     */
+    void call(ContextType context) const {
+        currentCalled = chainHead;
+
+        while(currentCalled) { 
+            currentCalled->call(context);
```
Comment (rgrover) [inline ble/CallChainOfFunctionPointersWithContext.h]: isn't this going to call chained function pointers recursively?

File: ble/FunctionPointerWithContext.h [hunk 7]
```diff
@@ -83,6 +95,31 @@ class FunctionPointerWithContext {
...
 
     /**
+     * @brief Same as above
+     */
+    void operator()(ContextType context) const {
+        call(context);
+    }
+
+    /** Same as above, workaround for mbed os FunctionPointer implementation. */
+    void call(ContextType context) {
```
Comment (rgrover) [inline ble/FunctionPointerWithContext.h]: it would be surprising to have call() behave differently depending on constness of the functionpointer object. Why do you need this?

File: ble/FunctionPointerWithContext.h [hunk 8]
```diff
@@ -83,6 +95,31 @@ class FunctionPointerWithContext {
...
+        _caller(this, context);
+    }
+
+    typedef void (FunctionPointerWithContext::*bool_type)() const;
+
+    /** 
+     * implementation of safe bool operator
+     */
+    operator bool_type() const {
```
Comment (rgrover) [inline ble/FunctionPointerWithContext.h]: why is this needed?

File: ble/Gap.h [hunk 9]
```diff
@@ -140,10 +140,15 @@ class Gap {
         return (durationInMillis * 1000) / UNIT_1_25_MS;
     }
 
+    typedef FunctionPointerWithContext<TimeoutSource_t> TimeoutEventCallback_t;
+    typedef CallChainOfFunctionPointersWithContext<TimeoutSource_t> TimeoutEventCallbackChain_t;
```
Comment (rgrover) [inline ble/Gap.h]: do we need a `TimeoutEventCallback_t` when we've got a `TimeoutEventCallbackChain_t`?

File: ble/Gap.h [hunk 10]
```diff
@@ -140,10 +140,15 @@ class Gap {
...
+
+    typedef FunctionPointerWithContext<const ConnectionCallbackParams_t *> ConnectionEventCallback_t;
+    typedef CallChainOfFunctionPointersWithContext<const ConnectionCallbackParams_t *> ConnectionEventCallbackChain_t;
```
Comment (rgrover) [inline ble/Gap.h]: do we need a callback_t when we've got a callbackChain_t?

File: ble/Gap.h [hunk 11]
```diff
@@ -140,10 +140,15 @@ class Gap {
...
+
+    typedef FunctionPointerWithContext<const DisconnectionCallbackParams_t*> DisconnectionEventCallback_t;
+    typedef CallChainOfFunctionPointersWithContext<const DisconnectionCallbackParams_t*> DisconnectionEventCallbackChain_t;
```
Comment (rgrover) [inline ble/Gap.h]: do we need a callback_t when we've got a callbackChain_t?

File: ble/Gap.h [hunk 12]
```diff
@@ -940,11 +959,14 @@ class Gap {
      */
     void onRadioNotification(void (*callback)(bool param)) {
         radioNotificationCallback.attach(callback);
+        // why does it start radio notification ? It is not even indicated in the
```
Comment (rgrover) [inline ble/Gap.h]: yes, this is un-necessary. Let's remove this.

File: ble/Gap.h [hunk 13]
```diff
@@ -940,11 +959,14 @@ class Gap {
...
     void onRadioNotification(void (*callback)(bool param)) {
         radioNotificationCallback.attach(callback);
+        // why does it start radio notification ? It is not even indicated in the 
+        // doc that it start the listening process
         initRadioNotification();
     }
     template <typename T>
     void onRadioNotification(T *tptr, void (T::*mptr)(bool)) {
         radioNotificationCallback.attach(tptr, mptr);
+        // why does it start radio notification ?
```
Comment (rgrover) [inline ble/Gap.h]: delete this

File: ble/GattClient.h [hunk 14]
```diff
@@ -244,7 +249,11 @@ class GattClient {
      * Set up a callback for read response events.
      */
     void onDataRead(ReadCallback_t callback) {
-        onDataReadCallback = callback;
+        onDataReadCallbackChain.add(callback);
+    }
+
+    ReadCallbackChain_t& onDataRead() {
```
Comment (rgrover) [inline ble/GattClient.h]: please add a documentation header. it might be good to add a code-example showing its use

File: ble/Gap.h [hunk 15]
```diff
@@ -893,7 +898,13 @@ class Gap {
      * Set up a callback for timeout events. Refer to TimeoutSource_t for
      * possible event types.
      */
-    void onTimeout(TimeoutEventCallback_t callback) {timeoutCallback = callback;}
+    void onTimeout(TimeoutEventCallback_t callback) {
+        timeoutCallbackChain.add(callback);
+    }
+
+    TimeoutEventCallbackChain_t& onTimeout() {
```
Comment (rgrover) [inline ble/Gap.h]: please add a documentation header. it might be good to add a code-example showing its use

File: ble/Gap.h [hunk 16]
```diff
@@ -903,6 +914,10 @@ class Gap {
     template<typename T>
     void onConnection(T *tptr, void (T::*mptr)(const ConnectionCallbackParams_t*)) {connectionCallChain.add(tptr, mptr);}
 
+    ConnectionEventCallbackChain_t& onconnection() {
```
Comment (rgrover) [inline ble/Gap.h]: please add a documentation header. it might be good to add a code-example showing its use

File: ble/Gap.h [hunk 17]
```diff
@@ -911,6 +926,10 @@ class Gap {
     template<typename T>
     void onDisconnection(T *tptr, void (T::*mptr)(const DisconnectionCallbackParams_t*)) {disconnectionCallChain.add(tptr, mptr);}
 
+    DisconnectionEventCallbackChain_t& onDisconnection() {
```
Comment (rgrover) [inline ble/Gap.h]: please add a documentation header. it might be good to add a code-example showing its use

File: ble/GattClient.h [hunk 18]
```diff
@@ -252,7 +261,11 @@ class GattClient {
      * @Note: Write commands (issued using writeWoResponse) don't generate a response.
      */
     void onDataWritten(WriteCallback_t callback) {
-        onDataWriteCallback = callback;
+        onDataWriteCallbackChain.add(callback);
+    }
+
+    WriteCallbackChain_t& onDataWritten() {
```
Comment (rgrover) [inline ble/GattClient.h]: please add a documentation header. it might be good to add a code-example showing its use

File: ble/GattClient.h [hunk 19]
```diff
@@ -281,7 +294,11 @@ class GattClient {
      * GATT server.
      */
     void onHVX(HVXCallback_t callback) {
-        onHVXCallback = callback;
+        onHVXCallbackChain.add(callback);
+    }
+
+    HVXCallbackChain_t& onHVX() {
```
Comment (rgrover) [inline ble/GattClient.h]: please add a documentation header. it might be good to add a code-example showing its use

File: ble/GattServer.h [hunk 20]
```diff
@@ -259,12 +275,16 @@ class GattServer {
...
      * some object.
      */
-    void onDataWritten(void (*callback)(const GattWriteCallbackParams *eventDataP)) {dataWrittenCallChain.add(callback);}
+    void onDataWritten(const DataWrittenCallback_t& callback) {dataWrittenCallChain.add(callback);}
     template <typename T>
     void onDataWritten(T *objPtr, void (T::*memberPtr)(const GattWriteCallbackParams *context)) {
         dataWrittenCallChain.add(objPtr, memberPtr);
     }
 
+    DataWrittenCallbackChain_t& onDataWritten() {
```
Comment (rgrover) [inline ble/GattServer.h]: please add a documentation header. it might be good to add a code-example showing its use

File: ble/GattServer.h [hunk 21]
```diff
@@ -302,6 +322,10 @@ class GattServer {
         return BLE_ERROR_NONE;
     }
 
+    DataReadCallbackChain_t& onDataRead() {
```
Comment (rgrover) [inline ble/GattServer.h]: please add a documentation header. it might be good to add a code-example showing its use

File: ble/CallChainOfFunctionPointersWithContext.h (hunk 3 above)
Comment (rgrover) [inline ble/CallChainOfFunctionPointersWithContext.h]: please add a documentation header. it might be good to add a code-example showing its use

File: ble/CallChainOfFunctionPointersWithContext.h [hunk 22]
```diff
@@ -120,11 +161,40 @@ class CallChainOfFunctionPointersWithContext {
...
+     * @brief same as above but with function call operator
+     */
+    void operator()(ContextType context) const {
+        call(context);
+    }
+
+    typedef void (CallChainOfFunctionPointersWithContext::*bool_type)() const;
+    void True() const {}
+
+    operator bool_type() const {
```
Comment (rgrover) [inline ble/CallChainOfFunctionPointersWithContext.h]: please add a documentation header. it might be good to add a code-example showing its use

File: ble/CallChainOfFunctionPointersWithContext.h [hunk 23]
```diff
@@ -139,6 +209,8 @@ class CallChainOfFunctionPointersWithContext {
 
 private:
     pFunctionPointerWithContext_t chainHead;
+    mutable pFunctionPointerWithContext_t currentCalled;
```
Comment (rgrover) [inline ble/CallChainOfFunctionPointersWithContext.h]: please add a comment why you'd like this to be mutable.

File: ble/FunctionPointerWithContext.h (hunk 8 above)
Comment (rgrover) [inline ble/FunctionPointerWithContext.h]: it might be good to add a code-example showing its use

File: source/DiscoveredCharacteristic.cpp [hunk 24]
```diff
@@ -31,6 +31,52 @@ DiscoveredCharacteristic::read(uint16_t offset) const
     return gattc->read(connHandle, valueHandle, offset);
 }
 
+struct OneShotReadCallback { 
+    static void launch(GattClient* client, Gap::Handle_t connHandle, 
+                       GattAttribute::Handle_t handle,const GattClient::ReadCallback_t& cb) {
```
Comment (rgrover) [inline source/DiscoveredCharacteristic.cpp]: add a space after the `,`

File: source/DiscoveredCharacteristic.cpp [hunk 25]
```diff
@@ -31,6 +31,52 @@ DiscoveredCharacteristic::read(uint16_t offset) const
     return gattc->read(connHandle, valueHandle, offset);
 }
 
+struct OneShotReadCallback { 
+    static void launch(GattClient* client, Gap::Handle_t connHandle, 
+                       GattAttribute::Handle_t handle,const GattClient::ReadCallback_t& cb) { 
+        OneShotReadCallback* oneShot = new OneShotReadCallback(client, connHandle, handle, cb);
+        oneShot->attach();
```
Comment (rgrover) [inline source/DiscoveredCharacteristic.cpp]: why is this empty call to attach() needed?

File: source/DiscoveredCharacteristic.cpp [hunk 26]
```diff
@@ -59,6 +105,52 @@ DiscoveredCharacteristic::writeWoResponse(uint16_t length, const uint8_t *value)
     return gattc->write(GattClient::GATT_OP_WRITE_CMD, connHandle, valueHandle, length, value);
 }
 
+struct OneShotWriteCallback { 
+    static void launch(GattClient* client, Gap::Handle_t connHandle, 
+                       GattAttribute::Handle_t handle,const GattClient::WriteCallback_t& cb) {
```
Comment (rgrover) [inline source/DiscoveredCharacteristic.cpp]: white space

**Comment by rgrover** [comment]
looks good other than the comments made already. This is a useful contribution.

File: ble/CallChainOfFunctionPointersWithContext.h (hunk 22 above)
Comment (rgrover) [inline ble/CallChainOfFunctionPointersWithContext.h]: why don't we just have `operator bool`?

File: ble/CallChainOfFunctionPointersWithContext.h (hunk 1 above)
Comment (pan-) [inline ble/CallChainOfFunctionPointersWithContext.h]: Oups sorry about this, I will fix it right away

File: ble/CallChainOfFunctionPointersWithContext.h (hunk 2 above)
Comment (pan-) [inline ble/CallChainOfFunctionPointersWithContext.h]: I will remove the note, the call is not recursive anymore

File: ble/CallChainOfFunctionPointersWithContext.h (hunk 6 above)
Comment (pan-) [inline ble/CallChainOfFunctionPointersWithContext.h]: No, it will call the const version of the function, unfortunately, we can't get rid of non const version because minar and mbed os does not allow to build a function pointer with a const member function. This defect, will be fixed, in the future

File: ble/FunctionPointerWithContext.h (hunk 7 above)
Comment (pan-) [inline ble/FunctionPointerWithContext.h]: it is a mistake, I don't know why I didn't save this change

File: ble/Gap.h (hunk 9 above)
Comment (pan-) [inline ble/Gap.h]: Yes, it is the kind of function embedded in the callchain: 

```c++
class Gap { 
    // ...
    void onTimeout(TimeoutEventCallback_t callback) {
        timeoutCallbackChain.add(callback);
    }
    // ...
};

struct MyObject { 
    void onTimeout(TimeoutSource_t source) { 
        (void) source;
    }
};

MyObject myObject;
Gap& gap = ble.gap();
gap.onTimeout(simpleFreeFunction);
gap.onTimeout(makeFunctionPointer(myObject, &MyObject::onTimeout));
```

File: ble/Gap.h (hunk 10 above)
Comment (pan-) [inline ble/Gap.h]: I think we do, callback_t will be used to register function pointer, callbackChain_t provide access to the callchain.

File: ble/CallChainOfFunctionPointersWithContext.h (hunk 22 above)
Comment (pan-) [inline ble/CallChainOfFunctionPointersWithContext.h]: operator bool open a hole in the type system, it allow the comparison of unrelated types if they implement this operator. 

```c++
struct A { 
    operator bool() { return true; }
};

struct B { 
    operator bool() { return true; } 
}

A a;
B b;

// this will compile!!!
if(a == b) { 
// we go there  ....
}
```

I will make an abstraction to generate safe_bool class, it will be more understandable

File: ble/CallChainOfFunctionPointersWithContext.h (hunk 23 above)
Comment (pan-) [inline ble/CallChainOfFunctionPointersWithContext.h]: Ok, I have added a comment in the code

File: source/DiscoveredCharacteristic.cpp (hunk 25 above)
Comment (pan-) [inline source/DiscoveredCharacteristic.cpp]: It will attach this object to the GattClient read callback chain

File: ble/CallChainOfFunctionPointersWithContext.h (hunk 22 above)
Comment (pan-) [inline ble/CallChainOfFunctionPointersWithContext.h]: You can read more information about it here: http://www.artima.com/cppsource/safebool.html 
In c++11, it is not useful because you can use `explicit operator bool`.

Comment (rgrover) [inline ble/CallChainOfFunctionPointersWithContext.h]: Vincent, thanks a lot for the explanation. Can we add something in the comments what the intention is?

if we're going through all this only to be able to simplify the call() operator, then I'd prefer that call() should internally (privately) determine `functionPointer != NULL`.

File: ble/Gap.h (hunk 10 above)
Comment (rgrover) [inline ble/Gap.h]: can't the callbackChain_t encapsulate the callback_t. I still don't understand why we don't have only the callbackChain_t.

File: ble/CallChainOfFunctionPointersWithContext.h (hunk 22 above)
Comment (pan-) [inline ble/CallChainOfFunctionPointersWithContext.h]: The conversion to bool is not about operator but function pointer semantic. I have replace some "true" C++ function pointer with `FunctionPointerWithContext` instance, the semantic should be equivalent otherwise it will break things.

File: ble/Gap.h (hunk 10 above)
Comment (pan-) [inline ble/Gap.h]: From my understanding, you want an inner type in callbackChain_t which define the type of the callback stored by the callback chain, is it what you want ? 

If so I have started to make the change but it make the declaration more complicated and exposes details of the implementation (the call chain type).

For instance, in Gap:

```c++
void onConnection(const ConnectionEventCallback_t& callback);
```

became 

```C++
    void onConnection(const ConnectionEventCallbackChain_t::FunctionPointerWithContext_t& callback);
```
And other part of the program (like BLE.h) which were using `Gap::ConnectionEventCallback_t` should now use `Gap::ConnectionEventCallbackChain_t::FunctionPointerWithContext_t` .

Is it what you want ?

Comment (rgrover) [inline ble/Gap.h]: i misread. my bad. nothing needs to be done here.

**Comment by rgrover** [comment]
this is a new feature. this pushes ble to v2.1.0
